xfrd.c

```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+			zone->soa_nsd_acquired = 0;
+			memset(&zone->soa_nsd, 0, sizeof(xfrd_soa_type));
+		}
 	}
+
+	/* discard any leading corrupt transfers. */
+	while(xfr && (xfr->state == xfrd_xfr_corrupt ||
+	              xfr->state == xfrd_xfr_inconsistent))
+	{
+		prev_xfr = xfr->prev;
+		xfrd_discard_zone_xfr(zone, xfr);
+		xfr = prev_xfr;
+	}
+
+	if(!xfr) {
+		/* reset soa_xfr information if no transfers are acquired to
+		   decide if the zone needs a refresh */
+		if((!zone->latest_xfr) ||
+		   (!zone->latest_xfr->prev && !zone->latest_xfr->acquired))
+		{
+			zone->soa_xfr_acquired = 0;
+			memset(&zone->soa_xfr, 0, sizeof(xfrd_soa_type));
+		}
+		if(committed) {
+			xfrd_schedule_zone_xfrs(
+				zone, zone->latest_xfr, committed);
+		}
+		return; /* continue */
+	} else if(!committed) {
+		assert(xfr->state == xfrd_xfr_ok ||
+		       xfr->state == xfrd_xfr_new);
+		if(xfr == zone->latest_xfr) {
+			zone->soa_xfr_acquired = xfr->acquired;
+			zone->soa_xfr = xfr->soa;
+		}
+		xfrd_schedule_zone_xfrs(zone, xfr, committed);
+		return; /* nothing to update yet, continue */
+	}
+
+	assert(committed);
+	assert(xfr && (xfr->state == xfrd_xfr_ok ||
+	               xfr->state == xfrd_xfr_new));
+
+	/* transfer committed to memory */
+	log_msg(LOG_INFO, "zone %s serial %u is updated to %u",
+		zone->apex_str,
+		ntohl(zone->soa_nsd.serial),
+		ntohl(xfr->soa.serial));
+	zone->soa_nsd = xfr->soa;
+	zone->soa_nsd_acquired = xfr->acquired;
+	xfrd->write_zonefile_needed = 1;
+
+	/* discard committed transfers */
+	while(xfr) {
+		prev_xfr = xfr->prev;
+		xfrd_discard_zone_xfr(zone, xfr);
+		xfr = prev_xfr;
+	}
+
+	if((!zone->latest_xfr) ||
+	   (!zone->latest_xfr->prev || !zone->latest_xfr->acquired))
+	{
+		zone->soa_xfr_acquired = 0;
+		memset(&zone->soa_xfr, 0, sizeof(xfrd_soa_type));
+	}
+
+	xfrd_schedule_zone_xfrs(zone, zone->latest_xfr, committed);
+	xfrd_send_notify(xfrd->notify_zones, zone->apex, &zone->soa_nsd);
 }
 
+/* zone transfers must be processed after all soa information returned by nsd
+   has been processed, it is unknown whether difffiles for successful transfers
+   can be discarded otherwise. */
 void
-xfrd_prepare_zones_for_reload()
+xfrd_process_xfrs(int committed)
 {
-	xfrd_zone_type* zone;
+	xfrd_zone_type *zone;
+
 	RBTREE_FOR(zone, xfrd_zone_type*, xfrd->zones)
 	{
-		/* zone has a disk soa, and no nsd soa or a different nsd soa */
-		if(zone->soa_disk_acquired != 0 &&
-			(zone->soa_nsd_acquired == 0 ||
-			zone->soa_disk.serial != zone->soa_nsd.serial))
+		time_t elapsed;
+
+		/* reloads are scheduled by xfrd_schedule_zone_xfrs if any
+		   transfers are added to the tasklist */
+		xfrd_process_zone_xfrs(zone, committed);
+
+		if(!committed) {
+			continue; /* continue if nothing has been committed */
+		}
+
+		if((zone->soa_notified_acquired != 0 &&
+		    compare_serial(ntohl(zone->soa_notified.serial),
+		                   ntohl(zone->soa_nsd.serial)) <= 0) ||
+		   (zone->soa_notified_acquired < zone->soa_nsd_acquired))
```

Comment:
This is a && b || c without brackets, what do you mean?  Also if a notify for a later serial arrives before you process the earlier serial, I would think that soa_notify_acquired (time_t) < nsd_acquired (time_t) but we do not want to just ignore that notification, instead act on it to fetch that serial number.

Reply:
Wait now I see (a&&b) brackets, so the comment is about the time comparison part and ignoring earlier notifications.